src/main/cpp/src/timezones.cu

```diff
@@ -118,6 +120,183 @@ auto convert_timestamp_tz(column_view const& input,
   return results;
 }
 
+template <typename timestamp_type>
+struct time_add_functor {
+  using duration_type = typename timestamp_type::duration;
+
+  lists_column_device_view const transitions;
+
+  size_type const tz_index;
+
+  int64_t const duration_scalar;
+
+  __device__ inline timestamp_type plus_with_tz(timestamp_type const& timestamp,
+                                                int64_t const& duration) const
+  {
+    if (duration == 0L) { return timestamp; }
+
+    auto const utc_instants = transitions.child().child(0);
+    auto const tz_instants  = transitions.child().child(1);
+    auto const utc_offsets  = transitions.child().child(2);
+
+    auto const epoch_seconds_utc = static_cast<int64_t>(
+      cuda::std::chrono::duration_cast<cudf::duration_s>(timestamp.time_since_epoch()).count());
+    // input epoch seconds
+
+    auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+    auto const list_size      = tz_transitions.size();
+
+    auto const transition_times_utc = cudf::device_span<int64_t const>(
+      utc_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    auto const transition_times_tz = cudf::device_span<int64_t const>(
+      tz_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    // step 1: Binary search on utc to find the correct offset to convert utc to local
+    auto const utc_it = thrust::upper_bound(
+      thrust::seq, transition_times_utc.begin(), transition_times_utc.end(), epoch_seconds_utc);
+    auto const utc_idx =
+      static_cast<size_type>(thrust::distance(transition_times_utc.begin(), utc_it));
+    auto const utc_list_offset = tz_transitions.element_offset(utc_idx - 1);
+    auto const to_local_offset =
+      static_cast<int64_t>(utc_offsets.element<int32_t>(utc_list_offset));
+
+    auto const to_local_offset_duration = cuda::std::chrono::duration_cast<duration_type>(
+      cudf::duration_s{static_cast<int64_t>(to_local_offset)});
+
+    auto const duration_typed = cuda::std::chrono::duration_cast<duration_type>(
+      cudf::duration_us{static_cast<int64_t>(duration)});
+
+    // step 2: add the duration to the local timestamp
+    auto const local_timestamp_res = timestamp + to_local_offset_duration + duration_typed;
+
+    auto const result_epoch_seconds = static_cast<int64_t>(
+      cuda::std::chrono::duration_cast<cudf::duration_s>(local_timestamp_res.time_since_epoch())
+        .count());
+
+    // step 3: Binary search on local to find the correct offset to convert local to utc
+    auto const local_it = thrust::upper_bound(
```

Comment:
Here too can we refactor the existing code so we can share most/all of it? If not can we list why and try to refactor the code to share as much as possible? If there isn't much that would overlap it is fine. But I still would like the code to be located near the other transition code with a name that is clear what it is doing differently.